torch/_inductor/scheduler.py

```diff
@@ -4145,6 +4186,102 @@ def resolve_pending_fusions(
 
                 fuse_two_nodes(node_key1, node_key2)
 
+        def evaluate_pending_template_fusions(
+            fusion_nodes: OrderedSet[BaseSchedulerNode],
+            seen_pair_speedup_fn: OrderedSet[Callable[[], bool]],
+        ) -> None:
+            """
+            Evaluate pending template fusions for a set of fusion candidate nodes.
+            fusion_type is either "epilogue" or "prologue" for logging purposes.
+            """
+            finished_fusions: OrderedSet[BaseSchedulerNode] = OrderedSet()
+
+            def fuse_if_speedup(node1, node2, speedup_fn):
+                if (
+                    self.can_fuse(node1, node2)
+                    and speedup_fn()
+                    and not self.will_fusion_create_cycle(node1, node2)
+                ):
+                    fuse_two_nodes(node1, node2)
+                    non_template_node = (
+                        node2 if is_epilogue_fusion(node1, node2) else node1
+                    )
+                    if non_template_node in pending_fusions:
+                        del pending_fusions[non_template_node]
+                    finished_fusions.add(non_template_node)
+
+            while len(fusion_nodes) > len(finished_fusions):
+                template_futures = []
+                future_to_pending_fusion = {}
+
+                remaining_nodes = fusion_nodes - finished_fusions
```

Comment:
This is fusion_nodes x finished_fusions complexity. anyway to do this differently? fusion_nodes is potentially large.